Symfony Messenger component documentation

sroze opened the pull request:
This pull-request adds some documentation for the Messenger component. As it is, it's extracted from the README built incrementally in the code pull-request.

I'm sure there will be lots to change when the implementation is merged :)

sroze committed:
Move the documentation from the main PR to RST format.

javiereguiluz committed:
Fixed the RST syntax issues

fabpot commented:
Should be updated as the component is now named Messenger.

wiese committed:
minor doc fixes for the Message component

sroze committed:
Merge pull request #1 from wiese/patch-3

minor doc fixes for the Message component

javiereguiluz commented on components/message.rst:
```diff
@@ -0,0 +1,284 @@
+=====================
+
+    The Message component helps application to send and receive messages
+    to/from other applications or via
```
This phrase looks unfinished. Also, it's not a very descriptive description. Should we mention at least "message bus" or some key concept so the readers know undoubtely what problem does this component solve? Thanks!

cafferata commented on components/message.rst:
```diff
@@ -0,0 +1,284 @@
+
+    $ composer require symfony/message
+
+Alternatively, you can clone the `<https://github.com/symfony/message>`_ repository.
```
This should be `https://github.com/symfony/messenger`.

cafferata commented on components/message.rst:
```diff
@@ -0,0 +1,284 @@
+
+.. code-block:: terminal
+
+    $ composer require symfony/message
```
This should be `symfony/messenger` conform the [composer.json](https://github.com/symfony/messenger/blob/master/composer.json).

mhujer commented on components/message.rst:
```diff
@@ -0,0 +1,284 @@
+libraries for now. You can use one of the following adapters:
+
+#. `PHP Enqueue bridge`_ to use one of their 10+ compatible queues such as
+   RabbitMq, Amazon SQS or Google Pub/Sub.
```
`RabbitMq` -> `RabbitMQ`

greg0ire commented on components/message.rst:
```diff
@@ -0,0 +1,284 @@
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+To allow us to receive and send messages on the same bus and prevent a loop, the
+message bus is equipped with the ``WrapIntoReceivedMessage`` received. It will
```
I'm having a hard time parsing this sentence, especially the last word :confused:

greg0ire commented on components/message.rst:
```diff
@@ -0,0 +1,284 @@
+A consumer is responsible for receiving messages from a source and dispatching
+them to the application.
+
+Let's say you already proceed some "orders" on your application using a
```
processed != proceed

greg0ire replied:
"in" your application?

greg0ire commented on components/message.rst:
```diff
@@ -0,0 +1,284 @@
+orders.
+
+You will read this CSV file and dispatch a ``NewOrder`` message. All you need to
+do is your custom CSV consumer and Symfony will do the rest.
```
"to do" => "to do is create your…" or "to write is your…

OskarStark replied:
consumer -> receiver?

greg0ire commented on components/message.rst:
```diff
@@ -0,0 +1,284 @@
+
+    use App\Message\MyMessage;
+
+    $result = $this->get('message_bus')->handle(new MyMessage(/* ... */));
```
Maybe use DI to get the `message_bus` service?

OskarStark commented on components/message.rst:
```diff
@@ -0,0 +1,284 @@
+.. code-block:: xml
+
+    <service id="App\Handler\MyMessageHandler">
+       <tag name="message_handler" />
```
should be `message.handler` according to `message.receiver` here: https://github.com/symfony/symfony-docs/pull/9437/files#diff-f4526c4dad3a0bd7e76fd35201a182f2R276

OskarStark commented on components/message.rst:
```diff
@@ -0,0 +1,284 @@
+       }
+    }
+
+.. code-block:: xml
```
you should add yaml config format, too

soyuka commented on components/message.rst:
```diff
@@ -0,0 +1,284 @@
+.. code-block:: yaml
+
+    framework:
+       message:
```
`messenger` here and below

andreybolonin commented on components/message.rst:
```diff
@@ -0,0 +1,284 @@
+Asynchronous messages
+~~~~~~~~~~~~~~~~~~~~~
+
+Using the Message Component is useful to decouple your application but it also
```
`Using the Messager Component`

OskarStark replied:
Messenger

andreybolonin commented on components/message.rst:
```diff
@@ -0,0 +1,284 @@
+Concepts
+--------
+
+.. image:: /_images/components/message/overview.png
```
image:: /_images/components/messager/overview.png

OskarStark replied:
messenger

andreybolonin commented on components/message.rst:
```diff
@@ -0,0 +1,284 @@
+.. code-block:: yaml
+
+    framework:
+       message:
```
`messager:`

OskarStark replied:
messenger

andreybolonin commented on components/message.rst:
```diff
@@ -0,0 +1,284 @@
+.. code-block:: yaml
+
+    framework:
+       message:
```
`messager:`

OskarStark replied:
messenger

soyuka commented on components/message.rst:
```diff
@@ -0,0 +1,284 @@
+.. note::
+
+    If the message cannot be guessed from the handler's type-hint, use the
+    ``handles`` attribute on the tag.
```
I think best would be an example. Also mention that the guessing is done on the `_invoke` method argument

OskarStark commented on components/message.rst:
```diff
@@ -0,0 +1,284 @@
+           arguments:
+               - "@serializer"
+               - "%new_orders_csv_file_path%"
+
```
please remove blank line here

OskarStark commented on components/message.rst:
```diff
@@ -0,0 +1,284 @@
+             arguments:
+                 - "@mailer"
+                 - "%to_email%"
+
```
please remove blank line here

OskarStark commented on components/message.rst:
```diff
@@ -0,0 +1,284 @@
+Your own receiver
+-----------------
+
+A consumer is responsible for receiving messages from a source and dispatching
```
consumer -> receiver ?

OskarStark commented on components/message.rst:
```diff
@@ -0,0 +1,284 @@
+           tags:
+               - message.receiver
+
+Finally, use your consumer:
```
consumer -> receiver?

OskarStark commented on components/message.rst:
```diff
@@ -0,0 +1,284 @@
+.. index::
+   single: Message
```
Messenger?

OskarStark commented on components/message.rst:
```diff
@@ -0,0 +1,284 @@
+.. index::
+   single: Message
+   single: Components; Message
```
Messenger?

JasonReading commented on components/message.rst:
```diff
@@ -0,0 +1,284 @@
+
+.. code-block:: terminal
+
+    $ bin/console message:consume App\MessageReceived\NewOrdersFromCsvFile
```
Typo here, should be `App\MessageReceiver\NewOrdersFromCsvFile`

sroze committed:
Update the documentation a bit, and add more FrameworkBundle documentation

OskarStark commented on components/messenger.rst:
```diff
@@ -0,0 +1,194 @@
+  use Symfony\Component\Messenger\HandlerLocator;
+  use Symfony\Component\Messenger\Middleware\HandleMessageMiddleware;
+
+    use App\Message\MyMessage;
```
```diff
-  use Symfony\Component\Messenger\MessageBus;
-  use Symfony\Component\Messenger\HandlerLocator;
-  use Symfony\Component\Messenger\Middleware\HandleMessageMiddleware;
-
-    use App\Message\MyMessage;
+    use App\Message\MyMessage;
+    use Symfony\Component\Messenger\MessageBus;
+    use Symfony\Component\Messenger\HandlerLocator;
+    use Symfony\Component\Messenger\Middleware\HandleMessageMiddleware;
```

sroze replied:
I'm not sure how it even ended like that 😂 Changed!

OskarStark commented on components/messenger.rst:
```diff
@@ -0,0 +1,194 @@
+    namespace App\MessageSender;
+
+    use Symfony\Component\Message\SenderInterface;
+    use App\Message\ImportantAction;
```
do we use ordered use statements in the docs? if yes, this must be above the `SenderInterface`-use

OskarStark commented on components/messenger.rst:
```diff
@@ -0,0 +1,194 @@
+    use Symfony\Component\Message\ReceiverInterface;
+    use Symfony\Component\Serializer\SerializerInterface;
+
+    use App\Message\NewOrder;
```
```diff
-    use Symfony\Component\Message\ReceiverInterface;
-    use Symfony\Component\Serializer\SerializerInterface;
-
-    use App\Message\NewOrder;
+    use App\Message\NewOrder;
+    use Symfony\Component\Message\ReceiverInterface;
+    use Symfony\Component\Serializer\SerializerInterface;
```

OskarStark commented:
Not sure its important, but in your PR header:

```diff
- This pull-request adds some documentation for the Message component. 
+ This pull-request adds some documentation for the Messenger component. 
```

sroze committed:
Uses `::` to display as code blocks

greg0ire commented on components/messenger.rst:
```diff
@@ -0,0 +1,192 @@
+The Messenger Component
+=======================
+
+    The Messenger component helps application send and receive messages to/from other applications or via message queues.
```
"an application" or "your application"

greg0ire commented on components/messenger.rst:
```diff
@@ -0,0 +1,192 @@
+Bus
+---
+
+The bus is used to dispatch messages. MessageBus' behavior is in its ordered
```
> MessageBus' behavior is in its ordered middleware stack.

This looks wrong and I don't know what to suggest because I'm not completely sure what you meant to say, and also, because it's the first time you mention `MessageBus`  in this document.

greg0ire commented on components/messenger.rst:
```diff
@@ -0,0 +1,192 @@
+    $bus = new MessageBus([
+        new HandleMessageMiddleware(new HandlerLocator([
+            MyMessage::class => $handler,
+        ]))
```
Missing comma at end of line

greg0ire commented on components/messenger.rst:
```diff
@@ -0,0 +1,192 @@
+---------------
+
+Using the ``SenderInterface``, you can easily create your own message sender.
+Let's say you already have an ``ImportantAction`` message going through the
```
Not sure what Symfony's policy is on this but I refrain to use elision in docs and would write "Let us say"

sroze committed:
Update typos and missing reference

sroze committed:
Finish the documentation about the custom adapter

sroze committed:
It helps multiple applications
 Please enter the commit message for your changes. Lines starting

sroze committed:
Fix unresolved reference

sroze committed:
Add Messenger in the topics

sroze committed:
Add a documentation about middlewares and update based on reviews

sroze commented:
Updated based on feedbacks, split the two documentations (i.e. as a "topic" and as a "component") and added more details. And Travis is green 🎉. 

Can I have the honor of another round of reviews?

javiereguiluz commented on components/messenger.rst:
```diff
@@ -0,0 +1,189 @@
+The bus is used to dispatch messages. The behaviour of the bus is in its ordered
+middleware stack. The component comes with a set of middlewares that you can use.
+
+When using the message bus with Symfony's FrameworkBundle, the following middlewares
```
`middleware` is uncountable in English, so no plural for it :)

sroze commented on messenger.rst:
```diff
@@ -0,0 +1,264 @@
+
+.. code-block:: xml
+
+    <service id="App\MessageHandler\MyMessageHandler">
```
This is proposed in the following PR: https://github.com/symfony/symfony/pull/26685. Though, it's controversial :)

yceruto commented on messenger.rst:
```diff
@@ -0,0 +1,264 @@
+
+Instead of calling a handler, you have the option to route your message(s) to a
+sender. Part of an adapter, it is responsible of sending your message somewhere.
+You can configuration which message is routed to which sender with the following
```
typo `configure`?

sroze committed:
Update based on comments

sroze commented:
Ready for another round of reviews 👍

boite commented on components/messenger.rst:
```diff
@@ -0,0 +1,189 @@
+The Messenger Component
+=======================
+
+    The Messenger component helps applications send and receive messages to/from other applications or via message queues.
```
```diff
-    The Messenger component helps applications send and receive messages to/from other applications or via message queues.
+    The Messenger component helps applications send messages to and receive messages from other applications and via message queues.
```

boite commented on components/messenger.rst:
```diff
@@ -0,0 +1,189 @@
+.. image:: /_images/components/messenger/overview.png
+
+**Sender**:
+   Responsible for serializing and sending the message to _something_. This
```
```diff
-   Responsible for serializing and sending the message to _something_. This
+   Responsible for serializing and sending messages to _something_. This
```

boite commented on components/messenger.rst:
```diff
@@ -0,0 +1,189 @@
+   something can be a message broker or a third party API for example.
+
+**Receiver**:
+   Responsible for deserializing and forwarding the messages to handler(s). This
```
```diff
-   Responsible for deserializing and forwarding the messages to handler(s). This
+   Responsible for deserializing and forwarding messages to handler(s). This
```

boite commented on components/messenger.rst:
```diff
@@ -0,0 +1,189 @@
+
+**Handler**:
+   Given a received message, contains the user business logic related to the
+   message. In practice, that is just a PHP callable.
```
How about:-

> Responsible for handling messages using the business logic applicable to the messages.

The part about PHP callable is confusing, explained better in the Handlers section, below and I think it should be omitted here.

boite commented on components/messenger.rst:
```diff
@@ -0,0 +1,189 @@
+
+.. note:
+
+    Every middleware need to implement the ``MiddlewareInterface`` interface.
```
need -> needs
or
need to -> must

boite commented on components/messenger.rst:
```diff
@@ -0,0 +1,189 @@
+Handlers
+--------
+
+Once dispatched to the bus, messages will be handled by a "message handler". A
```
"dispatched" is confusing here: according to the Bus section, the bus is responsible for dispatching messages; and the method used to get a message on to the bus is `handle()`.

sroze replied:
No, it's `dispatch` :)

boite commented on components/messenger.rst:
```diff
@@ -0,0 +1,189 @@
+Once dispatched to the bus, messages will be handled by a "message handler". A
+message handler is a PHP callable (i.e. a function or an instance of a class)
+that will do the required processing for your message. It _might_ return a
+result::
```
> It _might_ return a result

What happens if it does or does not return a result?  Does the following example code illustrate something about this statement?

boite commented on components/messenger.rst:
```diff
@@ -0,0 +1,189 @@
+--------
+
+The communication with queuing system or third parties is delegated to
+libraries for now.
```
It isn't at all clear what this means or what are the implications for someone who intends to communicate with a queuing system or third party. Also system -> systems. And drop "for now" unless you want to say something about what will happen in the future.

boite reviewed (changes requested):
Partial review.  Hope it helps.

Nyholm commented on components/messenger.rst:
```diff
@@ -0,0 +1,189 @@
+       }
+    }
+
+Same bus received and sender
```
Capitalize.

Nyholm commented on messenger.rst:
```diff
@@ -0,0 +1,263 @@
+        }
+    }
+
+Register your factory
```
`s/f/F/`

boite reviewed (changes requested):
Good docs!  Have managed to follow some and use the message bus and a message handler. :)

theofidry commented on components/messenger.rst:
```diff
@@ -0,0 +1,189 @@
+
+**Sender**:
+   Responsible for serializing and sending the message to _something_. This
+   something can be a message broker or a third party API for example.
```
`a messager handler, broker or a third-party API`.

Reasons:

- The current phrasing kinda implies the component is very queue oriented. It's not wrong but it's not limited to this use case
- Using it with a message handler is a case common enough to be mentioned IMO
- I'm not sure all the people are familiar with queues and its related vocabulary so putting `message handler` which is likely to be more commonly known will make it more accessible

theofidry commented on components/messenger.rst:
```diff
@@ -0,0 +1,189 @@
+--------
+
+Once dispatched to the bus, messages will be handled by a "message handler". A
+message handler is a PHP callable (i.e. a function or an instance of a class)
```
we could add a link to the official doc here: 
- callable: https://secure.php.net/manual/en/language.types.callable.php
- anonymous function: https://secure.php.net/manual/en/functions.anonymous.php
- an invokable class: https://secure.php.net/manual/en/language.oop5.magic.php#object.invoke

greg0ire commented on components/messenger.rst:
```diff
@@ -0,0 +1,189 @@
+           $this->filePath = $filePath;
+       }
+
+       public function receive() : \Generator
```
That's compatible

boite commented on messenger.rst:
```diff
@@ -0,0 +1,263 @@
+adapter. This will also configure the following for you:
+
+1. A ``messenger.default_sender`` sender to be used when routing messages
+2. A ``messenger.default_receiver`` receiver to be used when consuming messages.
```
These should be `messenger.sender.default` and `messenger.receiver.default`.

boite commented on messenger.rst:
```diff
@@ -0,0 +1,263 @@
+                'My\Message\Message':  messenger.default_sender # Or another sender service name
+
+Such configuration would only route the ``My\Message\Message`` message to be
+asynchronous, the rest of the messages would still be directly handled.
```
Also, routing to `messenger.default_sender` seems to be wrong:-

```
In CheckExceptionOnInvalidReferenceBehaviorPass.php line 32:
                                                                               
  [Symfony\Component\DependencyInjection\Exception\ServiceNotFoundException]   
  The service "messenger.asynchronous.middleware.send_message_to_producer" ha  
  s a dependency on a non-existent service "messenger.default_sender".
```
It should be `messenger.sender.default`.

sroze replied:
It's less clear with a single example IMHO. I've updated the sender configuration 👍

sroze commented on messenger.rst:
```diff
@@ -0,0 +1,263 @@
+
+    use Symfony\Bundle\FrameworkBundle\Controller\Controller;
+    use Symfony\Component\Messenger\MessageBusInterface;
+
```
Added 👍

sroze commented on messenger.rst:
```diff
@@ -0,0 +1,263 @@
+        messenger:
+            routing:
+                'My\Message\MessageAboutDoingOperationalWork': messenger.operations_sender
+                '*': messenger.default_sender
```
Yep, nice call. Updated 👍

sroze committed:
Update the documentation to reflect the latest changes

sroze commented on messenger.rst:
```diff
@@ -0,0 +1,263 @@
+
+Learn how to build your own adapters within the Component's documentation. Once
+you have built your classes, you can register your adapter factory to be able to
+use it via a DSN in the Symfony application.
```
In the "Use your adapter" section :)

sroze commented:
It's ready to be merged. Tweaks can follow in other PRs I suppose.

sroze committed:
Minor wording change

sroze commented:
@boite thank you very much for all your reviews, much appreciated! I've updated the wording of the sentence you highlighted in your last review.

boite commented:
Well done @sroze; this all looks good.

javiereguiluz commented:
Let's merge this because, otherwise, this would be one of those eternal pull requests that never get merged. We can improve contents later as needed. Thanks Samuel for your contribution and thanks to the community for the reviews!
